src/Features/LanguageServer/ProtocolUnitTests/Completion/CompletionResolveTests.cs

```diff
@@ -205,11 +252,32 @@ private static ClassifiedTextRun[] CreateClassifiedTextRunForClass(string classN
                 new ClassifiedTextRun("class name", className)
             };
 
-        private static async Task<LSP.CompletionList> RunGetCompletionsAsync(TestLspServer testLspServer, LSP.CompletionParams completionParams)
+        private static Task<LSP.CompletionList> RunGetCompletionsAsync(TestLspServer testLspServer, LSP.CompletionParams completionParams)
         {
             var clientCapabilities = new LSP.VSClientCapabilities { SupportsVisualStudioExtensions = true };
-            return await testLspServer.ExecuteRequestAsync<LSP.CompletionParams, LSP.CompletionList>(LSP.Methods.TextDocumentCompletionName,
+            return RunGetCompletionsAsync(testLspServer, completionParams, clientCapabilities);
+        }
+
+        private static async Task<LSP.CompletionList> RunGetCompletionsAsync(
+            TestLspServer testLspServer,
+            LSP.CompletionParams completionParams,
+            LSP.VSClientCapabilities clientCapabilities)
+        {
+            var completionList = await testLspServer.ExecuteRequestAsync<LSP.CompletionParams, LSP.CompletionList>(LSP.Methods.TextDocumentCompletionName,
                 completionParams, clientCapabilities, null, CancellationToken.None);
+
+            // Emulate client behavior of promoting "Data" completion list properties onto completion items.
+            if (clientCapabilities.HasCompletionListDataCapability() &&
+                completionList is VSCompletionList vsCompletionList &&
+                vsCompletionList.Data != null)
+            {
+                foreach (var completionItem in completionList.Items)
+                {
+                    completionItem.Data ??= vsCompletionList.Data;
```

Comment:
The server should never return a non-null value for data if there is data on the completion list right?  I think in that case the test should throw if completionItem.Data and vsCompletionList.Data are both non-null

Reply:
You can, if the completionItem.Data field is present then it overrides what the completion list provides in its Data field.